62.CAD/main.cpp

```diff
@@ -276,49 +291,746 @@ class CPolyline
+		// where it was initially of the form x^3 + ax^2 + bx + c = 0 and x was replaced by y - a/3
+		// so a/3 needs to be subtracted from the solution to the first form to get the actual solution
+		double p = ky - kx * kx;
+		double p3 = p * p * p;
+		double q = kx * (2.0 * kx * kx - 3.0 * ky) + kz;
+		double h = q * q + 4.0 * p3;
+
+		if (h >= 0.0)
+		{
+			h = sqrt(h);
+			double2 x = (double2(h, -h) - q) / 2.0;
+
+			// Solving Catastrophic Cancellation when h and q are close (when p is near 0)
+			if (abs(abs(h / q) - 1.0) < 0.0001)
+			{
+				// Approximation of x where h and q are close with no carastrophic cancellation
+				// Derivation (for curious minds) -> h=√(q²+4p³)=q·√(1+4p³/q²)=q·√(1+w)
+				   // Now let's do linear taylor expansion of √(1+x) to approximate √(1+w)
+				   // Taylor expansion at 0 -> f(x)=f(0)+f'(0)*(x) = 1+½x 
+				   // So √(1+w) can be approximated by 1+½w
+				   // Simplifying later and the fact that w=4p³/q will result in the following.
+				x = double2(p3 / q, -q - p3 / q);
+			}
+
+			double2 uv = 
+				double2(x.X > 0.0 ? 1.0 : x.X == 0.0 ? 0.0 : -1.0, x.Y > 0.0 ? 1.0 : x.Y == 0.0 ? 0.0 : -1.0) // TODO get sign(x) working
+				* double2(pow(abs(x.X), 1.0 / 3.0), pow(abs(x.Y), 1.0 / 3.0));
+			double t = uv.X + uv.Y - kx;
+			t = std::clamp(t, 0.0, 1.0);
+
+			// 1 root
+			double2 qos = c + (2.0 * a + b * t) * t;
+			res = double2(sqrt(dot(qos, qos)), t);
+		}
+		else
+		{
+			double z = sqrt(-p);
+			double v = acos(q / (p * z * 2.0)) / 3.0;
+			double m = cos(v);
+			double n = sin(v) * 1.732050808;
+			double t[3] = { m + m, -n - m, n - m };
+			
+			for (uint32_t i = 0; i < 3; i++)
+				t[i] = std::clamp(t[i] * z - kx, 0.0, 1.0);
+
+			// 3 roots
+			double2 qos = c + (2.0 * a + b * t[0]) * t[0];
+			double dis = dot(qos, qos);
+
+			res = double2(dis, t[0]);
+
+			qos = c + (2.0 * a + b * t[1]) * t[1];
+			dis = dot(qos, qos);
+			if (dis < res.X) res = double2(dis, t[0]);
+
+			qos = c + (2.0 * a + b * t[2]) * t[2];
+			dis = dot(qos, qos);
+			if (dis < res.X) res = double2(dis, t[2]);
+
+			res.X = sqrt(res.X);
+		}
+
+		return res.X;
+	}
+
+	class Segment 
+	{
+	public:
+		const QuadraticBezierInfo* originalBezier;
+		// because beziers are broken down,  depending on the type this is t_start or t_end
+		double t_start;
+		double t_end; // beziers get broken down
+
+		QuadraticBezierInfo getSplitCurve()
+		{
+			double2 a = originalBezier->p[0] - 2.0 * originalBezier->p[1] + originalBezier->p[2];
+			double2 b = 2.0 * (originalBezier->p[1] - originalBezier->p[0]);
+			double2 c = originalBezier->p[0];
+
+			return { 
+				a * (t_end - t_start) * (t_end - t_start),
+				(t_end - t_start) * (2 * a * t_start + b),
+				Hatch::evaluteBezier(originalBezier, t_start)
+			};
+		}
+
+		double intersect(const Segment& other) const
+		{
+			const std::array<double, 2> intersections = Hatch::intersectionAlgorithm(originalBezier, other.originalBezier);
```

Comment:
I see some inconsistency in here, this should return 4 doubles because we can have up to 4 real roots
If you've filtered out against the [0,1] range of the bezier which I suggest you do that outside the function, then why are you filtering again below? (which should be there)